File: app/javascript/genre_sliders/structures/doubly_linked_list/LinkedList.js
// Doubly Linked List

import Node from './Node';

const LinkedList = class {
  constructor() {
    this._length = 0;
    this.head = null;
    this.tail = null;
  }

  add = (value) => {
    const node = new Node(value);

    if (this._length > 0) {
      this.tail.next = node;
      node.previous = this.tail;
      this.tail = node;
    } else {
      this.head = node;
      this.tail = node;
    }

    this._length++;

    return node;
  }

  searchNodeAt = (position) => {
    let currentNode = this.head;
    const length = this._length;
    let count = 1;
    const message = { failure: 'Failure: non-existent node in this list.' };

    // 1st use-case: an invalid position
    if (length === 0 || position < 1 || position > length) {
      throw new Error(message.failure);
    }

    // 2nd use-case: a valid position
    while (count < position) {
      currentNode = currentNode.next;
      count++;
    }

    return currentNode;
  }

  remove = (position) => {
    let currentNode = this.head;
    const length = this._length;
    let count = 1;
    const message = {
      failure: 'Failure: non-existent node in this list.',
      success: 'Node was successfully removed.'
    };
    let beforeNodeToDelete = null;
    let nodeToDelete = null;
    let deletedNode = null;

    // 1st use-case: an invalid position
    if (length === 0 || position < 1 || position > length) {
      throw new Error(message.failure);
    }

    // 2nd use-case: the first node is removed
    if (position === 1) {
      this.head = currentNode.next;

      // there is a second node
      if (!this.head) {
        this.head.previous = null;
      // there is no second node
      } else {
        this.tail = null;
      }

    // 3rd use-case: the last node is removed
    } else if (position === this._length) {
      this.tail = this.tail.previous;
      this.tail.next = null;

    // 4th use-case: a middle node is removed
    } else {
      while (count < position) {
        currentNode = currentNode.next;
        count++;
      }

      beforeNodeToDelete = currentNode.previous;
      nodeToDelete = currentNode;
      afterNodeToDelete = currentNode.next;

      beforeNodeToDelete.next = afterNodeToDelete;
      afterNodeToDelete.previous = beforeNodeToDelete;
      deletedNode = nodeToDelete;
      nodeToDelete = null;
    }

    this._length--;

    return message.success;
  }
}

export default LinkedList;

File: app/services/linked_list.rb
# app/services/movies_list.rb

# Data structure for doubly linked list
class LinkedList
  attr_accessor :length, :head, :tail

  def initialize
    @length = 0
    @head = nil
    @tail = nil
  end

  def add(value)
    node = Node.new(value)

    if @length > 0
      @tail.next = node
      node.previous = @tail
    else
      @head = node
    end

    @tail = node
    @length += 1
  end

  def search_node_at(position)
    current_node = @head
    count = 1

    if @length.zero? || position < 1 || position > @length
      raise StandardError, 'Failure: non-existent node in this list'
    end

    while count < position
      current_node = current_node.next
      count += 1
    end

    current_node
  end

  def remove(position)
    current_node = @head
    count = 1

    # 1st use-case: an invalid position
    if @length.zero? || position < 1 || position > @length
      raise StandardError, 'Failure: non-existent node in this list'
    end

    # 2nd use-case: the first node is removed
    if position == 1
      @head = current_node.next

      # There is a second node
      if !@head
        @head.previous = nil

      # There is no second node
      else
        @tail = nil
      end

    # 3rd use-case: the last node is removed
    elsif position == @length
      @tail = @tail.previous
      @tail.next = nil

    # 4th use-case: a middle node is removed
    else
      while count < position
        current_node = current_node.next
        count += 1
      end

      before_node_to_delete = current_node.previous
      after_node_to_delete = current_node.next

      before_node_to_delete.next = after_node_to_delete
      after_node_to_delete.previous = before_node_to_delete

      @length -= 1

      Rails.logger.debug "\nNode was successfully removed\n"
    end
  end
end
